### PR TITLE
fix: add email_change support to generateLink

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
* Adds support to generate email links for changing a user's email

## What is the current behavior?
* `/admin/generate_link` endpoint does not support generating email links for changing emails 
* https://github.com/supabase/supabase/discussions/5125#discussioncomment-3172735
* Addresses #365 

## What is the new behavior?
* Introduces 2 new types for generating email links: `email_change_current`, `email_change_new` 
```bash

# Change email from foo@example.com --> bar@example.com

curl -X POST 'http://localhost:9999/admin/generate_link' 
-H 'Authorization: Bearer token' 
-d '{
  "type": "email_change_current", 
  "new_email": "bar@example.com", 
  "email": "foo@example.com"
}'

# Response
{
  "action_link": "http://localhost:9999/verify?token=813...2f8&type=email_change&redirect_to=http://localhost:3000",
  "email": "foo@example.com",
  "email_change_sent_at": "2022-07-22T12:49:45.906546+08:00",
  "email_confirmed_at": "2022-07-22T11:58:38.056107+08:00",
  "email_otp": "412620",
  "hashed_token": "813...2f8",
  "new_email": "bar@example.com",
  "verification_type": "email_change_current"
}

curl -X POST 'http://localhost:9999/admin/generate_link' 
-H 'Authorization: Bearer token' 
-d '{
  "type": "email_change_new", 
  "new_email": "bar@example.com", 
  "email": "foo@example.com"
}'

# Response {
  "action_link": "http://localhost:9999/verify?token=209...9de&type=email_change&redirect_to=http://localhost:3000",
  "email": "foo@example.com",
  "email_change_sent_at": "2022-07-22T12:49:45.906546+08:00",
  "email_confirmed_at": "2022-07-22T11:58:38.056107+08:00",
  "email_otp": "111222",
  "hashed_token": "209...9d3",
  "new_email": "bar@example.com",
  "verification_type": "email_change_new"
}
```
## Additional Context
* If `GOTRUE_MAILER_SECURE_EMAIL_CHANGE_ENABLED` is set to `false`, the following error message will be returned when attempting to generate an email link for the current email:

```bash
{
  "code": 422,
  "msg": "Enable secure email change to generate link for current email"
}
```

* If `GOTRUE_MAILER_SECURE_EMAIL_CHANGE_ENABLED` is set to `true`, then you will need to generate an email link using both `email_change_current` and `email_change_new` types.

### CODE REVIEW

File: api/mail.go [hunk 1]
```diff
@@ -52,13 +53,13 @@ func (a *API) GenerateLink(w http.ResponseWriter, r *http.Request) error {
...
-			if params.Type == "magiclink" {
-				params.Type = "signup"
+			if params.Type == magicLinkVerification {
+				params.Type = signupVerification
 				params.Password, err = password.Generate(64, 10, 0, false, true)
 				if err != nil {
 					return internalServerError("error creating user").WithInternalError(err)
 				}
-			} else if params.Type == "recovery" {
+			} else if params.Type == recoveryVerification || params.Type == "email_change_current" || params.Type == "email_change_new" {
```
Comment (J0) [inline api/mail.go:62]: /nit should we turn these into constants too

```suggestion
			} else if params.Type == recoveryVerification || params.Type == emailChangeCurrent || params.Type == emailChangeNew{
```

With corresponding constants defined somewhere

File: api/mail.go [hunk 2]
```diff
@@ -141,6 +142,28 @@ func (a *API) GenerateLink(w http.ResponseWriter, r *http.Request) error {
 			user.ConfirmationToken = hashedToken
 			user.ConfirmationSentAt = &now
 			terr = errors.Wrap(tx.UpdateOnly(user, "confirmation_token", "confirmation_sent_at"), "Database error updating user for confirmation")
+		case "email_change_current", "email_change_new":
```
Comment (J0) [inline api/mail.go:145]: similar to above

**Review by J0 — APPROVED** [review]
Looks good!

**Comment by BlockInside** [comment]
I'm making a request like this;
```js
const link_test = await supabase.auth.api.generateLink(
  "email_change_current",
  "x@x.com",
  {
    email: "x@x.com",
    new_email: "a@a.com"
  }
);

console.log('link_test: ', link_test);
```

And i'm getting this error;

```js
{
  data: null,
  error: { message: 'The new email address provided is invalid', status: 422 }
}
```

**Comment by 404Xavier** [comment]
### Hello folks is this change supported in the Supbase client libraries:

I am using Supabase in a project I am working on, I am getting weird errors when I try to generate the link using this API route what could be the problem?

```js
import supabaseServer from '../../lib/supabaseServer';

const handler = async (req, res) => {
	const resData = await supabaseServer.auth.api.generateLink(
		'email_change_current',
		'newemail@gmail.com'
	);
	console.log(resData);
	return res.status(200).send({
		message: 'It was successful'
	});
};

export default handler;
```
When I hit the above API route I get the following errors based on the email provided, both registered(current) and unregisted(new_email).
 - When I pass the `second argument` to the ` generateLink function` as my new email, I get this error
  ```js
    {
      "code": 404,
      "msg": "User not found"
    }
 ```
- When I pass the `second argument` to the `generateLink function` as my current email: I get the following error
 ```js
{
  "code": 422,
  "msg": "The new email address provided is invalid"
}
```
Seeing the above errors made me look at the `node_modules/@supabase/gotrue-js/src/lib/GoTrueApi.ts` and found the generateLink method which does not provide a way to provide the `new_email`  as an argument or an `option`. Which is the cause of the two weird errors seen above.

```js
async generateLink(
    type:
      | 'signup'
      | 'magiclink'
      | 'recovery'
      | 'invite'
      | 'email_change_current'
      | 'email_change_new',
    email: string,
    options: {
      password?: string
      data?: object
      redirectTo?: string
    } = {}
  ): Promise<{ data: Session | User | null; error: ApiError | null }> {
    try {
      const data: any = await post(
        this.fetch,
        `${this.url}/admin/generate_link`,
        {
          type,
          email,
          password: options.password,
          data: options.data,
          redirect_to: options.redirectTo,
        },
        { headers: this.headers }
      )
      return { data, error: null }
    } catch (e) {
      return { data: null, error: e as ApiError }
    }
  }
```
Having the above I tested the following code on the same API route and it worked seamlessly. Take note the example given above in the `What is new behavior` `cURL example` is incomplete as it does not provide an `apikey` in the header for the request to work successfully.
Below is the same using Axios library,
```js
import axios from 'axios';

//the request options
const  options = {
  method: 'POST',
  url: 'https://<project>.supabase.co/auth/v1/admin/generate_link',
  headers: {
    apikey: '<service_role_key_goes_here>',
    Authorization: 'Bearer <service_role_key_goes_here>',
  },
  data: {
    type: 'email_change_new', //or 'email_change_current' for the second link to be returned
    new_email: 'newemail@domain.foo',
    email: 'yourcurrentemail@domain.bar'
  }
};

axios.request(options).then(function (response) {
  console.log(response.data);
}).catch(function (error) {
  console.error(error);
});
```
I think the major problem is with the @supabase-js library which does not provide the mechanism to provide the second argument. If the above info is not clear please let me know. Thanks in advance.

**Pull request merged**